### ISSUE TITLE
make link syntax less clunky, more c++ like

### ISSUE THREAD

**Issue by jonathan-beard** [jonathan-beard/RaftLib#10]
<h3>Issue</h3>
I've had several requests to make the syntax for linking compute kernels less clunky and more C++ like. That means keeping the map::link syntax but also supporting something with a bit more polish. 

<h3>Potential Solution</h3>
```cpp
int
main( int argc, char **argv )
{  
   //generic random kernel instantiations
   kernel f;
   kernel g;
   //explicit declaration of map for example, potentially keep hidden one for compatibility
   map m;
   //in order version
   m >> f >> "a"_o >> "b"_i >> g;
   //out order version
   m >> f >> "a"_o >> raft::ooo >> "b"_i >> g;
   m.exe();
   return( EXIT_SUCCESS );
}
```

**Comment by adishavit** [comment]
A few questions:
 1. Can you give a reference to what graph this is building? Is it just a linear process?
 1. Maybe give the ubiquitous sum example?
 1. Also, let's try an example without an explicitly named kernel, i.e. one that is constructed and linked in-place.
 1. The `"a"_o` and `raft::ooo` are not clear. What are they doing? Is `"a"_o` even valid syntax?
 1. Ideally, since both `f` and `g` could be thought of as streams (input/filter/output streams) the explicit `map` seems redundant an unnatural (e.g. we don't write `std::ios << cout << 3 << endl`). Maybe make it so that the `map m` is implicit, for example, the in-order version would be:
```
auto& m = f >> "a"_o >> "b"_i >> g;
```
Or something like:
```
f >> "a"_o >> "b"_i >> g; // connect the graph
// one of...
f.run(); 
f.exe();
g.exe();
...
```
The  `operator <<` (and `operator >>`) will possibly return a `map` object and this may also be accessible through the streams' API too if you want the explicit handle.

**Comment by adishavit** [comment]
Here's I would imagine it:

This is the example from the paper:
```c++
const std::size_t count(100000);
auto linked_kernels(
   map.link(kernel::make<
      generate< std::int64_t > >(count),
      kernel::make<
      sum< std::int64_t,
      std::int64_t,
      std::int64_t > >(),
      "input_a"));
map.link(
   kernel::make< generate< std::int64_t > >(count),
   &(linked_kernels.dst),
   "input_b");
map.link(
   &(linked_kernels.dst),
   kernel::make< print< std::int64_t, ’\n’> >());
map.exe();
```

How about something like this (with named kernels):
```c++
const std::size_t count(100000);
auto src1    = kernel::make<generate< std::int64_t > >(count);
auto src2    = kernel::make<generate< std::int64_t > >(count);
auto summer  = kernel::make<sum< std::int64_t, std::int64_t, std::int64_t > >();
auto printer = kernel::make< print< std::int64_t, ’\n’> >();

// Option 1:
src1 >> summer["a"]; // generates the initial implicit map, links on "a" port
src2 >> summer["b"]; // links to the same map on "b" input port

// Option 2: Alternative syntax:
src1 >> summer; // generates the initial implicit map, links on first empty port of correct type
src2 >> summer; // links to the same map on the next empty port of the correct type

// Option 3: Alternative syntactic sugar:
src1 >> summer << src; // link to first and second ports

summer >> printer;        // or
summer["out"] >> printer; // or

// In many applications, the graph can already be running by now without an explicit call to exe(). Any unconnected nodes will just cause the graph to block until all nodes are in place.
// In fact, user placement can determine if anything actually happens.

// Otherwise, the map ref is available to all nodes, so one of:
src1.run();
src2.run();
summer.run();
printer.run();
```

**Comment by jonathan-beard** [comment]
```cpp
const auto count( 100000 );
using gen = generate< std::int64_t >;
using sm  = sum< std::int64_t, std::int64_t, std::int64_t >;
using pr    = print< std::int64_t, ’\n’>;

auto src1(  kernel::make< gen >( count ) );
auto src2(  kernel::make< gen >( count ) );
auto sum(  kernel::make< sm >() );
auto p( kernel::make< pr >() )

src1 >> "0"_o >> "input_a"_i >> sum;
src2 >> "0"_o >> "input_b"_i >> sum;
sum >> "out"_o >> "in"_i >> p;

map.exe();
```
So the 
```cpp
"input_a"_i
```
is the string literal syntax, so it's a function that creates an object in return. The idea behind this is that it gives the user a chance to use the same naming scheme as before, but in a much shorter fashion. 
The second path I'd thought of is this:
```cpp
src1( "0"_o ) >> sum( "input_a"_i );
```
where the port names and directions (either _o for output or _i for input) are explicit through the operator "()" overload of the kernel object.  I'll see if I can address some of the other thoughts in a bit.

**Comment by adishavit** [comment]
>
> const auto count( 100000 );using gen = generate< std::int64_t >;using sm  = sum< std::int64_t, std::int64_t, std::int64_t >;using pr    = print< std::int64_t, ’\n’>;
> auto src1(  kernel::make< gen >( count ) );auto src2(  kernel::make< gen >( count ) );auto sum(  kernel::make< sm >() );auto p( kernel::make< pr >() )
>
> So this just moves `kernel::make<>` into a separate definition, but it is
certainly clearer than the original (albeit the extra named variables).

> src1 >> "0"_o >> "input_a"_i >> sum;
> src2 >> "0"_o >> "input_b"_i >> sum;
> sum >> "out"_o >> "in"_i >> p;
>
> map.exe();
>
> So the
>
> "input_a"_i
>
> is the string literal syntax, so it's a function that creates an object in
> return. The idea behind this is that it gives the user a chance to use the
> same naming scheme as before, but in a much shorter fashion.
>

I don't know. The user-defined literal syntax here doesn't seem to fit with
the streaming concept/intuition. To me it just seems confusing.

Not sure what "0"_o is. Is "0" defined anywhere? Is it the default output
port of src1 ?


> The second path I'd thought of is this:
>
> src1( "0"_o ) >> sum( "input_a"_i );
>
> where the port names and directions (either _o for output or _i for input)
> are explicit through the operator "()" overload of the kernel object. I'll
> see if I can address some of the other thoughts in a bit.
>
This is closer to what I suggested, but:

   1. I prefer square brackets since that's how the internal string key is
   used inside the kernel;
   2. Why not just use the string as is and do the literal conversion
   implicitly/internally?


src1["0"] >> sum["input_a"]; // sum knows that "input_a" is an input port

src2["0"] >> sum["input_b"]; // sum knows that "input_a" is an input port

sum["out"] >> p["in"];

This seems very concise and clean to me stating the flow clearly and
minimally.

**Comment by jonathan-beard** [comment]
That's very strange on the markdown parsing. 

So for the example you've given:
```cpp
src1["0"] >> sum["input_a"]; // sum knows that "input_a" is an input port
src2["0"] >> sum["input_b"]; // sum knows that "input_a" is an input port
sum["out"] >> p["in"];
```

The issue I have with this from a syntactic point of view is that the user could (albeit stupidly) produce a bug where two ports are now named "input_a", one on the input and one on the output. Your original idea of using static look-up (i.e., declaring each port directly at compile time rather than dynamically would fix this issue) but it would also create other issues when the run-time tries to dynamically improve parallelize the application. 

I think something like:
```cpp
const auto count( 100000 );
using gen = generate< std::int64_t >;
using sm  = sum< std::int64_t, std::int64_t, std::int64_t >;
using pr    = print< std::int64_t, ’\n’>;

auto src1(  kernel::make< gen >( count ) );
auto src2(  kernel::make< gen >( count ) );
auto sum(  kernel::make< sm >() );
auto p( kernel::make< pr >() )

raft::map >> src1.output[ "0" ] >> sum.input[ "a" ];
raft::map >> src2.output[ "0" ] >> sum.input[ "b" ];
raft::map >> sum.output[ "out" ] >> p[ "in" ];

/** 
 * although as has been mentioned, we could start the kernels as soon as the
 * constructors call ret.
 */
raft::map.exe();
```

Though I still like the (perhaps more c++ stream-like syntax:
```cpp
src1 >> "0"_output >> "a"_input >> sum >> raft::map;
src2 >> "0"_output >> "b"_input >> sum >> raft::map;
sum >> "out"_output >> "in" >> p >> raft::map
```

I agree that the square brackets might end up being more "intuitive" given that the kernel uses the [] overload internally. It would also be far faster to implement, although I now have toy examples of both working. The code for the square brackets is also far easier to understand since there is less template code to slog through. 

Any more thoughts? I think I'll go with my first example with the square brackets. Perhaps moving the raft::map to the end of the >> since we're adding the whole assembly to the map. Makes sense, at least to me.

**Comment by adishavit** [comment]
>
> So for the example you've given:
>
> src1["0"] >> sum["input_a"]; // sum knows that "input_a" is an input port
> src2["0"] >> sum["input_b"]; // sum knows that "input_a" is an input port
> sum["out"] >> p["in"];
>
> The issue I have with this from a syntactic point of view is that the user
> could (albeit stupidly) produce a bug where two ports are now named
> "input_a", one on the input and one on the output.
>
Even so, this can be caught at initialization time - see below...

raft::map >> src1.output[ "0" ] >> sum.input[ "a" ];
> raft::map >> src2.output[ "0" ] >> sum.input[ "b" ];
> raft::map >> sum.output[ "out" ] >> p[ "in" ];
>
> Do you mean "0" as in the "first" port, or do you assume that the kernel
actually has an output port named by the arbitrary string "0"?

It feels to me that the "input/"output" specification is superfluous.

You can always ensure/enforce/warn on non-uniqueness, and require explicit
specification, as you propose, only when there is some ambiguity (this is
also in the spirit of C++).


> Though I still like the (perhaps more c++ stream-like syntax:
>
> src1 >> "0"_output >> "a"_input >> sum >> raft::map;
> src2 >> "0"_output >> "b"_input >> sum >> raft::map;
> sum >> "out"_output >> "in" >> p >> raft::map
>
> I agree that the square brackets might end up being more "intuitive" given
> that the kernel uses the [] overload internally. It would also be far
> faster to implement, although I now have toy examples of both working. The
> code for the square brackets is also far easier to understand since there
> is less template code to slog through.
>
This really feels unnatural and not only because of the unorthodox literal
syntax.
It is unnatural to stream labels and objects together.
I think of a kernel as an associative map to its ports.

> Any more thoughts? I think I'll go with my first example with the square
> brackets. Perhaps moving the raft::map to the end of the >> since we're
> adding the whole assembly to the map. Makes sense, at least to me.
>
As long as the map is required to be explicitly stated, I think that moving
it to the end is a wonderful improvement, which seems more intuitive and in
tune with the streaming concept.
I still think that it would be best if it could be omitted altogether in
most use cases (and exist and inferred implicitly).

**Comment by jonathan-beard** [comment]
Ok, here's what I think we can go with....which I think is keeping with C++ syntax (and it'll compile)
```cpp
int
main( int argc, char **argv )
{  

    kernel a;
    kernel b;
    
    map m;
    /** in order syntax, you can chain kernels **/ 
    a[ "A" ] >> b[ "B" ] >> a >>  m;
    /** out of order syntax, again..you can chain if you want **/
    a[ "A" ] >> raft::order::out >> b[ "B" ] >> m;
    /** split syntax, if nothing goes in between **/
    a >> raft::split( a[ " b " ], b, c, d ) >> raft::join( a,b,c,d ) >> m;
    /** 
     * split syntax if you want to have something in between the 
     * split and join. You can even define ports to split to as 
     * in a[ "foo" ] within the split as in raft::split( a[ "foo" ], b )
     * then connect follow on ports 
     */
    a >> raft::split( a, b, c, d ) >> m;
    
    /** 
     * we can keep this syntax as well so that the last >>
     * will return a link object with the source and destination
     * kernels if you want to create an inline, anonymous
     * string of kernels that should be linked somewhere
     * else.
     */
    auto link( a[ "A" ] >> b[ "B" ] >> a >>  m );

    return( EXIT_SUCCESS );
}
```

**Comment by adishavit** [comment]
Can you be more explicit about the graph you are building here?

    /** in order syntax, you can chain kernels **/
>     a[ "A" ] >> b[ "B" ] >> a >>  m;
>
> If `a[ "A" ]` is output port "A" on kernel `a`, what is "plain `a`?

>     /** out of order syntax, again..you can chain if you want **/
>     a[ "A" ] >> raft::order::out >> b[ "B" ] >> m;
>
> Not sure I get what this means...

>     /** split syntax, if nothing goes in between **/
>     a >> raft::split( a[ " b " ], b, c, d ) >> raft::join( a,b,c,d ) >> m;
>     /**      * split syntax if you want to have something in between the      * split and join. You can even define ports to split to as      * in a[ "foo" ] within the split as in raft::split( a[ "foo" ], b )     * then connect follow on ports      */
>     a >> raft::split( a, b, c, d ) >> m;
>
> I don't get what this does.


>     /**      * we can keep this syntax as well so that the last >>     * will return a link object with the source and destination     * kernels if you want to create an inline, anonymous     * string of kernels that should be linked somewhere     * else.     */
>     auto link( a[ "A" ] >> b[ "B" ] >> a >>  m );
>
>  This is nice.


I'm having second thoughts about having `m` on the end.
Looking at it now, it seems less natural because with iostreams, which is
the syntactic and conceptual model we are aiming for, the stream object,
e.g. `cout` or `cin` are always at the beginning.

So as long as `m` is explicitly mentioned, I think this syntax might look
more natural after all:

    m << a << b["B"] << a[ "A" ];

Or maybe, just support both streaming operators and let the user use
whatever is most intuitive or convenient.
In fact, we both operators + parentheses, you may be able to write complex
graphs with a single line:

m << (b >> c << a); // b->c, a->c, c->m

Would it be possible to eliminate `m` from the statement? Have it
internally passed around?

**Comment by jonathan-beard** [comment]
I'll address the rest of the comments shortly, but I want to see if I can figure out how to make this work b/c I like it. Not sure if it's intuitive or not, but definitely shorter:
```cpp
m << (b >> c << a); // b->c, a->c, c->m
```
I'm going to think out loud, well...silently via text:

I'm starting to realize that maybe we don't need the "src" kernel returned from the link statements. All the examples that users have sent me and in almost all the ones I've coded myself the src is never used. In this example it would be almost impossible to retrieve the src into kernel "c" (since there are two) simply. Returning only the destination might break some things, but not much. The only case would be if the user dynamically creates a source within a for loop and needs to capture the src., but the destination seems to be far more useful.

So, in short I'm not sure what to think about this yet. The lazy programmer side of me loves it, fewer keystrokes, more precise, etc. The part of me that realizes that we're going to have to teach people how to use this realizes that it might not be the most intuitive. In my experience, single chains are easier to understand, but I suppose they're still there, we're just enabling more advanced construction techniques.

**Comment by jonathan-beard** [comment]
So for the "split/join" cases:
```cpp
kernel a,b,c,d,e,f;
m <<  a << raft::split( b[ " B " ], c, d, e ) << raft::join( f );
```
The "manual" insertion of a split kernel takes care of situations where a library kernel exists, but we want to send the output to multiple other kernels. We have library kernels that can either distribute the data (passing out data as it comes in based on a scheduler) or copy the data to each kernel (i.e., sending the identical data to each of the kernels in the split). This eliminates the programmer from having to hack the library kernel "a" and enables them to optimize how it's being sent (i.e., copy, distribute, etc.). This is different from:

```cpp
m << (b >> c << a); 
```
Which would require a port to exist at run-time on "c" to collect the input from "a" and "b." I suppose "c" could be a split kernel allocated by the user just like the in-line construction above made the run-time create one. That would be both constructs would still be valid, i.e., the 
```cpp
raft::split( xxx );
```
syntax and the one you propose can exist side by side. I suppose both could be allowed since the split object exists, the number of ports is determined dynamically anyways so it wouldn't really matter if it's done using one construction or the other. Ok, so I think I'm going back to my original proposal, to have the map object on the left hand side and use the "<<" operator. I'll see if I can get the:
```cpp
m << (b >> c << a); 
```
syntax to work.

**Comment by adishavit** [comment]
You're right. I didn't quite think about the return value from the operators, and that it'd be context dependent... 

For << chaining, you want to return the rhs (operator right hand side) for the next item. You could call this upstream chaining.

For >> chaining, you also return the rhs. I guess his would be downstream chaining. 

However in mixed mode... It's a little less obvious... and more complicated than the single chains. 

Let me think out loud too:

Consider the following and assume both operators always return rhs:
`m << (b >> c) << a;`

This gives what we want: a fork. The parens are redundant. 

Basically chains do not need parens. They are left to right. 

Mixed operators are also left to right so might cause confusion. 

`m << b >> (c << a);`
This is equivalent to:

`m << c << a << b` 
Or 
`b >> a >> c >> m`

Because in the parens << returns rhs==`a`. 

Maybe there ought to be something like 
`m << b >> (c << a).sink();`
Where `.sink()` (or dst() or target()...) returns `c` in this case. 

Actually these issues do not need mixed operators. Consider:

`m << c << a << b` ==
`m << (c << a) << b; // c <- a <- b` 
Vs. 
`m << c << (a << b); // c <-b,  a <- b`
These build different graphs. 

But these issues might occur with any streaming operator in general. 


Also: would it be possible to get rid of `m` in the statement????

**Comment by adishavit** [comment]
So split and join are like mux/demux/duplicate operators via an `std::tie()` like interface?

**Comment by jonathan-beard** [comment]
I'm still thinking about the stream syntax, the map "m" was left in originally for mixing standard C/C++ code with the graph. i.e.:
```cpp
std::vector< int32_t > v;
kernel a( v ),b;
m << a << b
m.exe();
for( const auto val : v )
{
   /** do something **/
}
m << //create another graph structure here, but with implicit barrier at exe so those ops complete.
```

I agree that the for loop would be better served as another kernel, but I'm sure we can come up with some less contrived examples where we'd want to have a construct like this. Perhaps post processing? I suppose we could add in a barrier for a specific kernel instead of the more draconian approach that I took to get this project started.

**Comment by jonathan-beard** [comment]
>So split and join are like mux/demux/duplicate operators via an `std::tie()` like interface?
Yes.

**Comment by jonathan-beard** [comment]
so maybe a barrier is the solution. keep the "map" object "m" since it is equivalent to c++ stream. However, instead if the exe, have each kernel execute as soon as they're linked, at least in tandem. The question then becomes, when to check for a fully connected graph (loose edges indicate programmer error). currently non-connected edges are identified at the exe() step.

**Comment by jonathan-beard** [comment]
I think I have the best compromise solution:
```cpp
kernel a,b,c;
/**
 * create topology that looks like:
 * b->a
 * b->c
 */
raft::map += a << b >> c
```

This also enables the use of locally created sub-maps that can be added to the main map using the same graph syntax. The raft::map object has a global scope so adding to it within the operator overload can't easily (relative term I know) be turned off to add to a sub-map only. Now a user can make a sub-map object with defined ports. Add some internal configurations, say if they want to put together a jpeg encode pipeline, then plug it into bigger application with "<<" syntax to the main raft::map object. 

I also propose we get rid of the need to explicitly call exe (although I'll leave it for certain cases). The main raft::map can execute as soon as the main scope exits, or as soon as it reaches a barrier (basically a proxy for exe). The original reason for the "exe" command was so that we could check the graph before we start executing to ensure that it is fully connected (i.e. the user didn't leave dangling ports that might lead to an app never completing). A barrier seems like just as good a way, and perhaps more intuitive to our extant population of programmers. A raft::map::barrier( kernel &ref ) forms a barrier for the mixed sequential code (within main) to wait on a map (essentially force it's execution to that point)

**Comment by adishavit** [comment]
I like the use of += for adding to a map instead of the streaming operator.

As for `a <<  b >> c`, it would normally be called as in `((a << b) >> c)`.
For me, this means stream b->a, then stream the result to c. 
The question is whether or not this "result" is `(b->a) -> c`, treating (b->a) as a combined link/partial path, or, somewhat less intuitively, return `b` again for parallel streaming to `c`. I think the first is the more intuitive.

This is different from `a >> b << c` == `((a >> b) << c)` which would actually seem to create `c -> a -> b`.
None of these would actually seem to create a->b, c->b or b->a, b->c.

I think the guiding rule needs to be that a streaming operator retunes the result of the linkage, i.e. a partial path. Note that a single kernel is a partial path of length 0. 
Then, with left to right evaluation order (or parens):
 - >> streams the path output to the next partial path /kernel input, and;
 - << streams the next partial path output into the current partial path *input*.

Thus, I think I am coming to the following conclusion:
 1. I am not sure that we need both operators. `<<` should suffice.
 2. Streaming operators should only used for linear chaining.

If you want demuxing, define multiple chains as in

```
m += a << b;
m += c << b;
```
or 
```
m += a << b["even"];
m += c << b["odd"];
```

(I apologize for any confusion, I know I suggested some of these ideas)

**Comment by jonathan-beard** [comment]
Hmm.....now I'll have to think about that a bit more. I still like the idea of chaining in both directions...but you're right. It is confusing if you're not very careful. 

So, any thoughts on the barrier vs. +=. I'm thinking of going to waitForKernel( kernel &ref ) vs. barrier. Seems both might be intuitive, but for different crowds.

**Comment by adishavit** [comment]
As I said, I like the += instead of calling `m << ...`.
I don't really get the barrier stuff.
As I see it, unless explicitly turned off, the graph can start streaming as soon as connections are made.
This is due to the following reasons:

 - If the graph is not fully connected it will just hang/block until the it is connected further downstream.
 - With a single source (or kernels requiring all sources), the user can always decide to connect the source nodes last, thus preventing anything from happening.

If `m` goes out of scope, what's left?

**Comment by jonathan-beard** [comment]
how about the case where you want to interact with something in main at two
different places. i.e., a std::vector from one kernel and another further
down within the main function. A barrier would enable procedural code to
pull in data from that kernel within main, then move on down the line.
Without that, how does the user know that the vector insert is complete?

**Comment by adishavit** [comment]
I still don't understand, can you give a pseudocode example?
But, as I said, this can be the default, but you can still give the used manual control over when things start to run.

**Comment by jonathan-beard** [comment]
Sure:
```cpp
int main( int argc, char **argv )
{
std::vector< int > v;
std::vector< float > f;
kernel a,b( v ),c,d( f );
raft::map += a >> b >> c >> d;
raft::map::barrier( b );
for( const auto val : v )
{
   //do something important
}
/** do some more stuff **/
//random code
//random code
raft::map::barrier( d )
for( const auto val : f )
{
  //do something else important
}
return( EXIT_SUCCESS );
}
```

**Comment by jonathan-beard** [comment]
So the final rev 1.x syntax (I've gotta get this finalized at least) is:
```cpp
/** given a few kernels and the move operator **/
kernel b,d;
/** the += returns a link to the last kernel mapped in the chain **/
auto &dst( raft::map += kernel( "a" ) >> b >> kernel( "c" );
raft::map += dst >> d;
```

**Comment by adishavit** [comment]
This example looks good.
A few clarifications though:
- For op `>>` a lhs kernel will mean its output port (e.g. `kernel( "a" )`), and the rhs means the kernel's input port (e.g.`kernel( "c" )`).
 - The opposite hold for `<<`, if it is implemented.
 - In you example, the whole statement returns the last kernel, so essentially the object created by `kernel( "c" )`.
 - The more I think about it, the more I think that supporting both operators will cause more confusion and ambiguity, so for rev 1.x I suggest avoiding one of them. The main reason that it isn't when to return the head of a path and when the tail.

**Comment by jonathan-beard** [comment]
So, I'm going to get a bit more detailed:
```cpp
kernel& operator >> ( kernel &src, kernel &dst )
{
 /** connection op **/
 return( dst );
}
```
So the LHS of this op is the destination. Likewise if I implement it (I'm sticking to a linear arrangement for now) the other direction (I think for the moment the left to right construction is more "intuitive" since many languages are left to right, although I think the alternate construction might be better for right to left natives).

I think eventually we'll derive a better way to derive the link path, I'm of the same opinion....it's currently ambiguous and doesn't really buy us much in the way of terseness.

**Comment by adishavit** [comment]
Short answer: Yes, the code is what I had in mind. Return `dst`. 
I think you have a type: You meant *RHS* being the destination for operator `>>`. LHS is the source.
I think not implementing op `<<` for now is fine.

Longer answer: Strictly speaking, there is another option which is: not to return either `src` or `dst` but to return some 3rd type, call it *partial path*, which has a head and a tail that can be appended to.In this case head would be `src` and the tail would be `dst`.

**Comment by jonathan-beard** [comment]
wow, can't believe it's been a month since the last post on this. time flies when you're writing papers, coding, and traveling. I'm about to commit the final updates for this syntax...and I say final because I put it in the camera ready journal paper. one addition to the above is adding a double bracket syntax so that 

```cpp
raft::kernel a;
raft::kernel c;
raft::kernel d;
raft::map += ( a >> b[ "in" ][ "out" ] >> c );
```

is now possible...still working out the details. In the journal version I didn't mention the second brackets only that the middle kernels could  be multiple in/single out. This extends that so that kernels can be multiple in, multiple out with the ">>" syntax. 

Any thoughts?

**Comment by adishavit** [comment]
I like this.
If only we could have left hand side brackets!
